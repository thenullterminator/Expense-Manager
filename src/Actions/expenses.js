import uuid from 'uuid';

// Action Generators for expense reducers.
const addExpense=({description='',note='',amount=0,createdAt=0}={})=>{

      return {
                   type:'ADD_EXPENSE',
                   expense: {
                        id:uuid(),
                        description,
                        note,
                        amount,
                        createdAt
             }
       };
 };
 
 const removeExpense=({id}={})=>{
 
       return{
                   type:'REMOVE_EXPENSE',
                   id
        };
 };
 
 const editExpense=(id,updates)=>{
       return{
             type:'EDIT_EXPENSE',
             id,
             updates
       };
 
 };

export {addExpense,removeExpense,editExpense};