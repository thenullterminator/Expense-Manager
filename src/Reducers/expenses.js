// Expense Reducer
const expenseDefaultState=[];
const expenseReducer=(state=expenseDefaultState,action)=>{

      switch(action.type)
      {
            case 'ADD_EXPENSE':
                  return [...state,action.expense];
            case 'REMOVE_EXPENSE':
                  return state.filter(({id})=>id!==action.id);
            case 'EDIT_EXPENSE' :
                  return state.map((expense)=>{
                        if(expense.id===action.id)
                        {
                              return{
                                    ...expense,
                                    ...action.updates
                              };
                        }
                        else
                        {
                              return expense;
                        }
                  });
            default:
                  return state;
      }

};

export default expenseReducer;