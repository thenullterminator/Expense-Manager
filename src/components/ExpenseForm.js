import React from 'react';
import moment from 'moment';
import ModernDatepicker from 'react-modern-datepicker';

console.log(moment().valueOf());
export default class ExpenseForm extends React.Component{

      state={
            description:'',
            note:'',
            amount:'',
            createdAt:moment(),
            error:''
      };

      onDescriptionChange=(e)=>{
            this.setState({
                  description:e.target.value
            });
      };

      onNoteChange=(e)=>{
            this.setState({
                  note:e.target.value
            });
      };

      onAmountChange=(e)=>{

            const amount=e.target.value;
            if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
            {
                  this.setState({
                        amount
                  });
            }
            
      };

      onDateChange=(date)=>{
            if(date)
            {
                  this.setState({createdAt:date});     
            }
      };

      onSubmitForm=(e)=>{

            e.preventDefault();
            if(!this.state.description || !this.state.amount)
            {
                  this.setState({error:'Please provide both Description and Amount.'});
            }
            else
            {
                  this.setState({error:''});

                  this.props.onSubmit({
                        description:this.state.description,
                        amount:parseFloat(this.state.amount)*100,
                        createdAt:this.state.createdAt.valueOf(),
                        note:this.state.note
                  });
            }
      };

      render(){
            return (
                  <div>
                        {this.state.error && <p>{this.state.error}</p>}
                        <form onSubmit={this.onSubmitForm}>
                              <input
                                    type='text'
                                    placeholder='Description'
                                    autoFocus
                                    value={this.state.description}
                                    onChange={this.onDescriptionChange}
                              />
                              <input
                                    type='text'
                                    placeholder='Amount'
                                    value={this.state.amount}
                                    onChange={this.onAmountChange}
                              />
                              <p></p>
                              <textarea 
                                    placeholder='Add a note for the expense'
                                    value={this.state.note}
                                    onChange={this.onNoteChange}
                              ></textarea>

                              <ModernDatepicker
                                    date={this.state.createdAt.format('DD-MM-YYYY')}
                                    format={'DD-MM-YYYY'}
                                    showBorder
                                    onChange={this.onDateChange}
                                    placeholder={'Select a date'}
                              />
                              <button>Add Expense</button>
                        </form>
                  </div>
            );
      };
}