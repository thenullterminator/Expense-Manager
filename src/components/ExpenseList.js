import React from  'react';
import {connect} from 'react-redux';

import selectExpenses from '../selectors/selectors';
import ExpenseListItem from './ExpenseListItem'

const ExpenseList=(props)=>(

      <div>
            <p>This is expense list</p>
            {props.expenses.map((expense,index)=>(<ExpenseListItem key={index} {...expense} />))}
      </div>

);

const mapStateToProps=(state)=>{

      return {
            expenses:selectExpenses(state.expenses,state.filters)
      };
};

export default connect(mapStateToProps)(ExpenseList);