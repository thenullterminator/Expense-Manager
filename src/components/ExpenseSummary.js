import React from 'react';
import selectExpenseTotal from '../selectors/expense-total';
import selectExpenses from '../selectors/selectors';
import {connect} from 'react-redux';
import numeral from 'numeral';

export const ExpenseSummary=(props)=>{

      return(

            <div>
            <p>Viewing {props.expenses.length} expense(s) totalling {numeral(selectExpenseTotal(props.expenses)).format('$0,0.00')}</p>
            </div>

      );

};

const mapStateToProps=(state)=>{

      return {
            expenses:selectExpenses(state.expenses,state.filters)
      };
};

export default connect(mapStateToProps)(ExpenseSummary);