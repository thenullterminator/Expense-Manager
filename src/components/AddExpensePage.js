import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import {addExpense} from '../Actions/expenses';


const AddExpensePage=(props)=>(
      <div>
            <h2>Add Expense</h2>
            <ExpenseForm onSubmit={(expense)=>{
                  props.dispatch(addExpense(expense));
                  props.history.push('/');//Redirecting to home page.
            }}/>
      </div>
);

export default connect()(AddExpensePage);
