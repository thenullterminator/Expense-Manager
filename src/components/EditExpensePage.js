import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../Actions/expenses';

export class EditExpensePage extends React.Component{

      onSubmit=(expense)=>{
            this.props.editExpense(this.props.expense.id,expense);
            this.props.history.push('/');//Redirecting to home page.
      };

      onRemove=(e)=>{
            this.props.removeExpense(this.props.expense.id);
            this.props.history.push('/');//Redirecting to home page.
      };

      render(){
            return (
                  <div>
                        <ExpenseForm
                              expense={this.props.expense}
                              onSubmit={this.onSubmit}
                        />

                        <button onClick={this.onRemove}>Remove</button>
                  </div>
            );
      }
};

const mapStateToProps=(state,props)=>{
      return{
            expense:state.expenses.filter((expense)=> (expense.id===props.match.params.id))[0]
      };
};

const mapDispatchToProps=(dispatch,props)=>({
      editExpense:(id,expense) => dispatch(editExpense(id,expense)),
      removeExpense:(id) => dispatch(removeExpense({id}))
});

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage);
