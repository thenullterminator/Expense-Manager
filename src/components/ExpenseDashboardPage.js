import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummary from '../components/ExpenseSummary';
const ExpenseDashboardPage=()=>(
      <div>
            <p>This is expense dashboard page.</p>
            <ExpenseSummary/>
            <ExpenseListFilter/>
            <ExpenseList/>
      </div>
);

export default ExpenseDashboardPage;
