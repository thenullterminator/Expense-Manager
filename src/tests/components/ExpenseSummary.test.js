import React from 'react';
import {shallow} from 'enzyme';
import {ExpenseSummary} from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render Expense Summary correctly for no expenses',()=>{

      const wrapper=shallow(<ExpenseSummary expenses={[]}/>);
      expect(wrapper).toMatchSnapshot();
});

test('should render Expense Summary correctly for multiple expenses',()=>{

      const wrapper=shallow(<ExpenseSummary expenses={expenses}/>);
      expect(wrapper).toMatchSnapshot();
});

