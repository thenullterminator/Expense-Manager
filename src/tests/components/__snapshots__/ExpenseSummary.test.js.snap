// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense Summary correctly for multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     expense(s) totalling 
    $39.00
  </p>
</div>
`;

exports[`should render Expense Summary correctly for no expenses 1`] = `
<div>
  <p>
    Viewing 
    0
     expense(s) totalling 
    $0.00
  </p>
</div>
`;
